pkg/3scale/amp/operator/imagestream_reconciler.go

```diff
@@ -4,137 +4,105 @@ import (
 				r.Logger().Error(createErr, fmt.Sprintf("Error creating object %s. Requeuing request...", objectInfo))
 				return createErr
 			}
-			r.Logger().Info(fmt.Sprintf("Created object %s", objectInfo))
 			return nil
 		}
 		return err
 	}
 
-	needsUpdate, err := r.ensureImageStream(existingImageStream, desiredImageStream)
-	if err != nil {
-		return err
-	}
-
-	if needsUpdate {
-		r.Logger().Info(fmt.Sprintf("Updating ImageStream %s", objectInfo))
-		err := r.Client().Update(context.TODO(), existingImageStream)
-		if err != nil {
-			r.Logger().Error(err, fmt.Sprintf("error updating ImageStream %s", objectInfo))
-			return err
-		}
+	if r.reconciler.IsUpdateNeeded(desiredIS, existingIS) {
+		return r.updateResource(existingIS)
 	}
 
 	return nil
 }
 
-func (r *ImageStreamReconciler) ensureImageStream(updated, desired *imagev1.ImageStream) (bool, error) {
-	changed := false
-
-	objectMetaChanged, err := r.ObjectMetaMerger.EnsureObjectMeta(&updated.ObjectMeta, &desired.ObjectMeta)
-	if err != nil {
-		return false, err
-	}
-	if objectMetaChanged {
-		changed = true
-	}
-
-	r.ensureImageTagReferences(updated, desired)
-	if !reflect.DeepEqual(updated.Spec, desired.Spec) {
-		updated.Spec = desired.Spec
-		changed = true
-	}
-
-	return changed, nil
+type ImageStreamGenericReconciler struct {
 }
 
-// Sets the generation field of the desired ImageStreamTags with the value
-// of the equivalent ImageStreamTag in the existing ImageStreamTag
-// That's because the Generation field in the ImageStream TagReferences
-// is filled by OpenShift after deploying so that would
-// make the comparison on that field always with a result of being
-// unequal when comparing the generation field between existing and
-// desired
-// It also sets the ReferencePolicyType in the desired in case it is empty
-// because OpenShift fills it with a value when not defined
-// The arrays are sorted because there could be the same tags
-// but in different order and the comparison should be performed
-// independently of the order of the arrays
-func (r *ImageStreamReconciler) ensureImageTagReferences(updated, desired *imagev1.ImageStream) {
-	updatedImageStreamTagReferenceMap := map[string]*imagev1.TagReference{}
-	for idx := range updated.Spec.Tags {
-		tagref := &updated.Spec.Tags[idx]
-		updatedImageStreamTagReferenceMap[tagref.Name] = tagref
-	}
-
-	for idx := range desired.Spec.Tags {
-		desiredTagRef := &desired.Spec.Tags[idx]
-		if updatedTagRef, ok := updatedImageStreamTagReferenceMap[desiredTagRef.Name]; ok {
-			desiredTagRef.Generation = updatedTagRef.Generation
+func NewImageStreamGenericReconciler() *ImageStreamGenericReconciler {
+	return &ImageStreamGenericReconciler{}
+}
 
-			if desiredTagRef.ReferencePolicy.Type == "" {
-				desiredTagRef.ReferencePolicy.Type = updatedTagRef.ReferencePolicy.Type
+func (r *ImageStreamGenericReconciler) IsUpdateNeeded(desired, existing *imagev1.ImageStream) bool {
+	// merging approach will be implemented
```

Comment:
Is there a specific motivation for this in front on just defining what's in the "desired" ImageStream?

Reply:
IMO, both approaches are valid for imagestreams (not for all k8s resources). Merging approach would leave existing tags while updating approach would delete existing tags not desired.

If there is no clear reason to implement "updating" approach, I would leave it as it is now, i.e. merging approach

Reply:
:+1: